src/Control/Monad/Catch.hs

```diff
@@ -179,6 +178,27 @@ class MonadCatch m => MonadMask m where
   -- and/or unkillable.
   uninterruptibleMask :: ((forall a. m a -> m a) -> m b) -> m b
 
+  -- | A generalized version of the standard bracket function which allows
+  -- distinguishing different exit cases. Instead of providing it a single
+  -- cleanup action, this function takes two different actions: one for the
+  -- case of a successful run of the inner function, and one in the case of an
+  -- exception. The former function is provided the acquired value and the
+  -- inner function's result, and returns a new result value. The exception
+  -- cleanup function is provided both the acquired value and the exception
+  -- that was thrown.
+  --
+  -- @since 0.9.0
+  generalBracket
+    :: m a
+    -- ^ acquire some resource
+    -> (a -> b -> m c)
+    -- ^ cleanup, no exception thrown
+    -> (a -> SomeException -> m ignored)
+    -- ^ cleanup, some exception thrown. The exception will be rethrown
```

Comment:
To me at least, this looks odd: not using complete sentences, but still having a period and a capital to start the next pseudo-sentence.

Maybe use a semi-colon instead?